substrate/frame/contracts/src/benchmarking/mod.rs

```diff
@@ -468,19 +468,36 @@ benchmarks! {
 	// It creates a maximum number of metering blocks per byte.
 	// `c`: Size of the code in bytes.
 	#[pov_mode = Measured]
-	upload_code {
+	upload_code_determinism_enforced {
 		let c in 0 .. T::MaxCodeLen::get();
 		let caller = whitelisted_caller();
 		T::Currency::set_balance(&caller, caller_funding::<T>());
-		let WasmModule { code, hash, .. } = WasmModule::<T>::sized(c, Location::Call);
+		let WasmModule { code, hash, .. } = WasmModule::<T>::sized(c, Location::Call, false);
 		let origin = RawOrigin::Signed(caller.clone());
-	}: _(origin, code, None, Determinism::Enforced)
+	}: upload_code(origin, code, None, Determinism::Enforced)
 	verify {
 		// uploading the code reserves some balance in the callers account
 		assert!(T::Currency::total_balance_on_hold(&caller) > 0u32.into());
 		assert!(<Contract<T>>::code_exists(&hash));
 	}
 
+	// Uploading code with [`Determinism::Relaxed`] should be more expensive than uploading code with [`Determinism::Enforced`],
+	// as we always try to save the code with [`Determinism::Enforced`] first.
+	#[pov_mode = Measured]
+	upload_code_determinism_relaxed {
+		let c in 0 .. T::MaxCodeLen::get();
+		let caller = whitelisted_caller();
+		T::Currency::set_balance(&caller, caller_funding::<T>());
+		let WasmModule { code, hash, .. } = WasmModule::<T>::sized(c, Location::Call, true);
+		let origin = RawOrigin::Signed(caller.clone());
+	}: upload_code(origin, code, None, Determinism::Relaxed)
+	verify {
+		assert!(T::Currency::total_balance_on_hold(&caller) > 0u32.into());
```

Comment:
Here I see assertion on two cases, however there's also the case of `Determinism::Relaxed` but no in-deterministic code, in which case there shouldn't be any penalty? Could be worthwhile to check unless this is already asserted on in some other place.

Reply:
this just assert here that we did indeed benchmark the most expensive code path.
since we will first try to load the module in deterministic mode, then fallback to non-deterministic.
The assertion here just validate that. All scenarios are tested in src/tests.rs below